synapse/storage/data_stores/main/devices.py

```diff
@@ -172,16 +211,40 @@ def get_devices_by_remote(self, destination, from_stream_id, limit):
         # means that there are more than limit updates all of which have the same
         # steam_id.
 
+        # figure out which cross-signing keys were changed by intersecting the
+        # update list with the master/self-signing key by user maps
+        cross_signing_keys_by_user = {}
+        for user_id, device_id, stream, _opentracing_context in updates:
+            if device_id == master_key_by_user.get(user_id, {}).get("pubkey", None):
+                result = cross_signing_keys_by_user.setdefault(user_id, {})
+                result["master_key"] = master_key_by_user[user_id]["key_info"]
```

Comment:
can we not do this in the previous loop, rather than iterating over `updates` twice?